### PR TITLE
[News]: Extract the logic from fetching feeds into a reusable class

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/31747

This is in preparation for introducing a new, alternative feed controller for an experiment we're doing with the Brave News feed. The change is internal only, and should not affect what's displayed in the feed.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/feed_controller.cc [hunk 1]
```diff
@@ -180,24 +152,16 @@ void FeedController::EnsureFeedIsUpdating() {
...
-            base::Unretained(controller), std::move(publishers));
-        // Perform all feed downloads in parallel
-        auto fetch_items_handler =
-            base::BarrierCallback<FeedItems>(2, std::move(feed_items_handler));
-        controller->FetchCombinedFeed(fetch_items_handler);
-        VLOG(1) << "Feed Controller found " << direct_feed_publishers.size()
-                << " direct feeds.";
-        controller->direct_feed_controller_->DownloadAllContent(
-            std::move(direct_feed_publishers), fetch_items_handler);
+            base::Unretained(controller), std::move(publishers)));
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_controller.cc:162]: Happy to just change these to `WeakPtrs` :smile:

File: components/brave_news/browser/feed_fetcher.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+namespace brave_news {
+
+using FeedItems = std::vector<mojom::FeedItemPtr>;
+using ETags = base::flat_map<std::string, std::string>;
+using FetchFeedCallback = base::OnceCallback<void(FeedItems items, ETags tags)>;
+using UpdateAvailableCallback = base::OnceCallback<void(bool)>;
+
+class FeedFetcher {
+ public:
+  explicit FeedFetcher(
```
Comment (simonhong) [inline components/brave_news/browser/feed_fetcher.h:31]: only need `explicit` for ctor with single parameter.

```suggestion
  FeedFetcher(
```

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: components/brave_news/browser/feed_fetcher.h (hunk 2 above)
Comment (fallaciousreasoning) [reply components/brave_news/browser/feed_fetcher.h]: Nice catch!

**Pull request merged**